server/sonar-server/src/main/java/org/sonar/server/measure/ws/ComponentTreeDataLoader.java

```diff
@@ -0,0 +1,390 @@
+        }
+      });
+      Set<String> missingMetricKeys = Sets.difference(
+        new LinkedHashSet<>(request.getMetricKeys()),
+        new LinkedHashSet<>(foundMetricKeys));
+
+      throw new NotFoundException(format("The following metric keys are not found: %s", Joiner.on(", ").join(missingMetricKeys)));
+    }
+
+    return metrics;
+  }
+
+  private Table<String, MetricDto, MeasureDto> searchMeasuresByComponentUuidAndMetric(DbSession dbSession, ComponentTreeData.Builder data) {
+    List<ComponentDtoWithSnapshotId> components = data.getComponentsFromDb();
+    List<MetricDto> metrics = data.getMetrics();
+
+    Map<Long, ComponentDtoWithSnapshotId> componentsBySnapshotId = Maps.uniqueIndex(components, ComponentDtoWithSnapshotIdToSnapshotIdFunction.INSTANCE);
+
+    Map<Integer, MetricDto> metricsById = Maps.uniqueIndex(metrics, MetricDtoFunctions.toId());
+    List<MeasureDto> measureDtos = dbClient.measureDao().selectBySnapshotIdsAndMetricIds(dbSession,
+      new ArrayList<>(componentsBySnapshotId.keySet()),
+      new ArrayList<>(metricsById.keySet()));
+
+    Table<String, MetricDto, MeasureDto> measuresByComponentUuidAndMetric = HashBasedTable.create(components.size(), metrics.size());
+    for (MeasureDto measureDto : measureDtos) {
+      measuresByComponentUuidAndMetric.put(
+        componentsBySnapshotId.get(measureDto.getSnapshotId()).uuid(),
+        metricsById.get(measureDto.getMetricId()),
+        measureDto);
+    }
+
+    addBestValuesToMeasures(measuresByComponentUuidAndMetric, data);
+
+    return measuresByComponentUuidAndMetric;
+  }
+
+  private static List<WsMeasures.Period> periodsFromSnapshot(SnapshotDto baseSnapshot) {
+    List<WsMeasures.Period> periods = new ArrayList<>();
+    for (int periodIndex = 1; periodIndex <= 5; periodIndex++) {
+      if (baseSnapshot.getPeriodDate(periodIndex) != null) {
+        periods.add(snapshotDtoToWsPeriod(baseSnapshot, periodIndex));
+      }
+    }
+
+    return periods;
+  }
+
+  private static WsMeasures.Period snapshotDtoToWsPeriod(SnapshotDto snapshot, int periodIndex) {
+    WsMeasures.Period.Builder period = WsMeasures.Period.newBuilder();
+    period.setIndex(periodIndex);
+    if (snapshot.getPeriodMode(periodIndex) != null) {
+      period.setMode(snapshot.getPeriodMode(periodIndex));
+    }
+    if (snapshot.getPeriodModeParameter(periodIndex) != null) {
+      period.setParameter(snapshot.getPeriodModeParameter(periodIndex));
+    }
+    if (snapshot.getPeriodDate(periodIndex) != null) {
+      period.setDate(formatDateTime(snapshot.getPeriodDate(periodIndex)));
+    }
+
+    return period.build();
+  }
+
+  /**
+   * Conditions for best value measure:
+   * <ul>
+   *   <li>component is a production file or test file</li>
+   *   <li>metric is optimized for best value</li>
+   * </ul>
+   */
+  private static void addBestValuesToMeasures(Table<String, MetricDto, MeasureDto> measuresByComponentUuidAndMetric, ComponentTreeData.Builder data) {
```

Comment:
Optimization - no need to filter components then loop over them at line 221 if `metricDtosWithBestValueMeasure` is empty.